spinalcordtoolbox/moco.py

```diff
@@ -0,0 +1,894 @@
+        if im_data.orientation[2] in 'LR':
+            cmd = ['isct_antsRegistration',
+                   '-d', '2',
+                   '--transform', 'Affine[%s]' %param.gradStep,
+                   '--metric', param.metric + '[' + file_dest + ',' + file_src + ',1,' + metric_radius + ',Regular,' + param.sampling + ']',
+                   '--convergence', param.iter,
+                   '--shrink-factors', '1',
+                   '--smoothing-sigmas', param.smooth,
+                   '--verbose', '1',
+                   '--output', '[' + file_mat + ',' + file_out_concat + ']']
+            cmd += sct.get_interpolation('isct_antsRegistration', param.interp)
+            if im_mask is not None:
+                # if user specified a mask, make sure there are non-null voxels in the image before running the registration
+                if np.count_nonzero(im_mask.data):
+                    cmd += ['--masks', im_mask.absolutepath]
+                else:
+                    # Mask only contains zeros. Copying the image instead of estimating registration.
+                    sct.copy(file_src, file_out_concat, verbose=0)
+                    do_registration = False
+                    # TODO: create affine mat file with identity, in case used by -g 2
+        # 3D mode
+        else:
+            cmd = ['isct_antsSliceRegularizedRegistration',
+                   '--polydegree', param.poly,
+                   '--transform', 'Translation[%s]' %param.gradStep,
+                   '--metric', param.metric + '[' + file_dest + ',' + file_src + ',1,' + metric_radius + ',Regular,' + param.sampling + ']',
+                   '--iterations', param.iter,
+                   '--shrinkFactors', '1',
+                   '--smoothingSigmas', param.smooth,
+                   '--verbose', '1',
+                   '--output', '[' + file_mat + ',' + file_out_concat + ']']
+            cmd += sct.get_interpolation('isct_antsSliceRegularizedRegistration', param.interp)
+            if im_mask is not None:
+                cmd += ['--mask', im_mask.absolutepath]
+        # run command
+        if do_registration:
+            kw.update(dict(is_sct_binary=True))
+            env = dict()
+            env.update(os.environ)
+            env = kw.get("env", env)
+            # reducing the number of CPU used for moco (see issue #201)
+            env["ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS"] = "1"
+            status, output = sct.run(cmd, verbose=1 if param.verbose == 2 else 0, **kw)
+
+    elif param.todo == 'apply':
+        sct_apply_transfo.main(args=['-i', file_src,
+                                     '-d', file_dest,
+                                     '-w', file_mat + param.suffix_mat,
+                                     '-o', file_out_concat,
+                                     '-x', param.interp,
+                                     '-v', '0'])
+
+    # check if output file exists
```

Comment:
We should possibly check the return value of sct_apply_transfo instead since there are other ways that this can fail (e.g. running out of memory when writing the file, etc.)

Reply:
`sct_apply_transfo` was only one of the if case, and `file_out_concat` can also be generated by the commands `isct_antsSliceRegularizedRegistration` (3d) or `isct_antsRegistration` (2d) as you can see in the other if cases. 

In the past i've tried to catch non-zero output from ANTs function, but in some OSs, the function can fail while outputing zero. 

So as a pragmatic approach, I decided to go with the "output file checking" approach, which is 100% sensitive.

Reply:
I've noticed this too so I check both now.